dali/pipeline/graph/op_graph2.cc

```diff
@@ -0,0 +1,313 @@
+// Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/pipeline/graph/op_graph2.h"
+#include <string>
+#include <utility>
+#include "dali/core/error_handling.h"
+#include "dali/pipeline/operator/name_utils.h"
+
+namespace dali {
+namespace graph {
+
+namespace {
+
+OpType ParseOpType(const std::string &device) {
+  if (device == "gpu") {
+    return OpType::GPU;
+  } else if (device == "cpu") {
+    return OpType::CPU;
+  } else if (device == "mixed") {
+    return OpType::MIXED;
+  }
+  DALI_FAIL("Unsupported device type: " + device + ".");
+}
+
+StorageDevice ParseStorageDevice(const std::string &io_device) {
+  if (io_device == "cpu") {
+    return StorageDevice::CPU;
+  }
+  return StorageDevice::GPU;
+}
+
+}  // namespace
+
+//////////////////////////////////////////////////////////////////////////////
+// OpGraph
+
+OpNode &OpGraph::AddOp(std::string instance_name, OpSpec spec) {
+  OpNodeList tmp;
+  auto &op_node = tmp.emplace_back(std::move(instance_name), std::move(spec));
+  if (!name2op_.emplace(op_node.instance_name, &op_node).second) {
+    throw std::invalid_argument(
+        make_string("Duplicate operator instance name: \"", op_node.instance_name, "\""));
+  }
+  op_node.iter = tmp.begin();
+  op_nodes_.splice(op_nodes_.end(), tmp);
+  return op_node;
+}
+
+DataNode &OpGraph::AddData(std::string name, StorageDevice device) {
+  DataNodeList tmp;
+  auto &data_node = tmp.emplace_back(std::move(name), device);
+  if (!name2data_.emplace(data_node.name, &data_node).second) {
+    throw std::invalid_argument(
+        make_string("Duplicate data node name: \"", data_node.name, "\""));
+  }
+  data_node.iter = tmp.begin();
+  data_nodes_.splice(data_nodes_.end(), tmp);
+  return data_node;
+}
+
+void OpGraph::RemoveDataNodeReferences(OpNode &op) {
+  for (auto *data : op.inputs) {
+    auto it = std::remove_if(data->consumers.begin(), data->consumers.end(), [&](DataEdge e) {
```

Comment:
Would you care to explain? `remove_if` moves the "removed" elements to the end and returns an iterator to the first removed element. `remove_if` has no power to modify the collection, because it only knows iterators. Having said that - I wonder what it does with `std::list`...